services/pgnotifier/pgnotifier.go

```diff
@@ -0,0 +1,447 @@
+			var count int
+			err := notifier.dbHandle.QueryRow(stmt).Scan(&count)
+			if err != nil {
+				logger.Errorf("PgNotifier: Failed to query for tracking jobs by batch_id: %s, connInfo: %s", stmt, notifier.URI)
+				panic(err)
+			}
+			if count == 0 {
+				stmt = fmt.Sprintf(`SELECT payload, status FROM %s WHERE batch_id = '%s'`, queueName, batchID)
+				rows, err := notifier.dbHandle.Query(stmt)
+				if err != nil {
+					panic(err)
+				}
+				defer rows.Close()
+				responses := []ResponseT{}
+				for rows.Next() {
+					var payload json.RawMessage
+					var status string
+					err = rows.Scan(&payload, &status)
+					responses = append(responses, ResponseT{
+						Status:  status,
+						Payload: payload,
+					})
+				}
+				*ch <- responses
+				break
+			} else {
+				logger.Infof("PgNotifier: Pending %d files to process in batch: %s", count, batchID)
+			}
+		}
+	})
+}
+
+func (notifier *PgNotifierT) updateClaimedEvent(id int64, tx *sql.Tx, ch chan ClaimResponseT) {
+	rruntime.Go(func() {
+		response := <-ch
+		var err error
+		if response.Err != nil {
+			stmt := fmt.Sprintf(`UPDATE %[1]s SET status=(CASE
+					WHEN	attempt > %[2]d THEN CAST ( '%[3]s' AS pg_notifier_status_type)
+				ELSE  CAST( '%[4]s' AS pg_notifier_status_type)
+				END), updated_at = '%[5]s', error = '%[6]s' WHERE id = %[7]v`, queueName, maxAttempt, AbortedState, FailedState, GetCurrentSQLTimestamp(), response.Err.Error(), id)
+			_, err = tx.Exec(stmt)
+		} else {
+			stmt := fmt.Sprintf(`UPDATE %[1]s SET status='%[2]s', updated_at = '%[3]s', payload = '%[4]s' WHERE id = %[5]v`, queueName, SucceededState, GetCurrentSQLTimestamp(), response.Payload, id)
+			_, err = tx.Exec(stmt)
+		}
+
+		if err == nil {
+			tx.Commit()
+		} else {
+			// TODO: verify rollback is necessary on error
+			tx.Rollback()
+			logger.Errorf("PgNotifier: Failed to update claimed event: %v", err)
+		}
+	})
+}
+
+func (notifier *PgNotifierT) Claim(workerID string) (claim ClaimT, claimed bool) {
+	//Begin Transaction
+	tx, err := notifier.dbHandle.Begin()
+	if err != nil {
+		return
+	}
+
+	var claimedID int64
+	var batchID, status string
+	var payload json.RawMessage
+	// Dont panic if acquire fails -- Just rollback & return the worker to free state again
+	stmt := fmt.Sprintf(`UPDATE %[1]s SET status='%[2]s',
+						updated_at = '%[3]s',
+						last_exec_time = '%[3]s',
+						worker_id = '%[4]v'
+						WHERE id = (
+						SELECT id
+						FROM %[1]s
+						WHERE status='%[5]s' OR status='%[6]s'
+						ORDER BY id
+						FOR UPDATE SKIP LOCKED
+						LIMIT 1
+						)
+						RETURNING id, batch_id, status, payload;`, queueName, ExecutingState, GetCurrentSQLTimestamp(), workerID, WaitingState, FailedState)
+	err = tx.QueryRow(stmt).Scan(&claimedID, &batchID, &status, &payload)
+
+	if err != nil {
+		logger.Errorf("PgNotifier: Claim failed: %v, query: %s, connInfo: %s", err, stmt, notifier.URI)
+		// TODO: verify rollback is necessary on error
+		tx.Rollback()
```

Comment:
Rollback is not necessary, but tx.Commit seems to be necessary - from other references in JobsDB.